pricegraph/src/graph/shortest_paths.rs

```diff
@@ -117,25 +199,55 @@ where
         }
     }
 
+    fn prepare_next_relaxation_step(&mut self) {
+        if let PredecessorStore::Bounded(predecessors_at_step, distance) =
+            &mut self.predecessor_store
+        {
+            predecessors_at_step.push(vec![None; self.graph.node_bound()]);
+            distance.current = distance.update.clone();
+        }
+    }
+
     /// Checks for negative weight cycle and, if any is found, creates loop in
     /// the predecessor store
     fn mark_cycle(&mut self) -> Option<G::NodeId> {
-        // This is the last step of the Bellman-Ford algorithm. It tries to relax
-        // each node: if a node can be relaxed then a negative cycle exists and is
-        // created in the predecessor store; otherwise no such cycle exists.
-
-        for i in self.graph.node_identifiers() {
-            for edge in self.graph.edges(i) {
-                let j = edge.target();
-                let w = *edge.weight();
-
-                if *self.distance(i) + w < *self.distance(j) {
-                    self.update_predecessor(j, Some(i));
-                    return Some(j);
+        match &mut self.predecessor_store {
+            PredecessorStore::Unbounded(..) => {
+                // This is the last step of the Bellman-Ford algorithm. It tries to relax
+                // each node: if a node can be relaxed then a negative cycle exists and is
+                // created in the predecessor store; otherwise no such cycle exists.
+                for i in self.graph.node_identifiers() {
+                    for edge in self.graph.edges(i) {
+                        let j = edge.target();
+                        let w = *edge.weight();
+                        if *self.distance(i) + w < *self.distance(j) {
+                            self.update_predecessor(j, Some(i));
+                            return Some(j);
+                        }
+                    }
+                }
+                None
+            }
+            PredecessorStore::Bounded(predecessors_at_step, _) => {
+                let steps = predecessors_at_step.len();
+                for end_node in self.graph.node_identifiers() {
+                    let mut node = end_node;
+                    for step in (0..steps).rev() {
```

Comment:
Can you give an example of where we have a cycle in some intermediate step but not in the last one? I'm asking to see if we could just store the last predecessor vector rather than having one for every step.

Reply:
This should be a unit test I think.

Reply:
I'm not sure what you mean by "intermediate step".
If you mean that it finds a loop when `step != 0` then this should be already covered in the test `bounded_search_with_negative_cycle_not_involving_origin`.
If you mean a test where the cycle doesn't start with `end_node`, this should be impossible: `self.graph.node_identifiers()` should contain all nodes in the graph, and if there is a cycle then a node in the cycle will eventually be selected as the `end_node`, which means that the cycle occurs at the last step.

I wonder if there would be an efficiency improvement if I used `self.connected_nodes()` here instead of `self.graph.node_identifiers()`.

Reply:
I meant an example for when it finds a loop with `n` steps, but not with `max_steps` or `max_steps-1` steps. I was under the impression that any cycle that exists in the `n`-th predecessor list would also exist in the last/last but one predecessor list.

Reply:
Not with the current implementation, at least. Example:
```
    #[test]
    fn bounded_search_with_negative_cycle_not_involving_origin_modified() {
        //   0 --1.0-> 5 --1.0-> 6 --1.0-> 1 --1.0-> 2
        //                                 ∧         |
        //                              -100.0      1.0
        //                                 |         v
        //                                 4 <-1.0-- 3
        let graph = Graph::<(), f64>::from_edges(&[
            (0, 5, 1.0),
            (5, 6, 1.0),
            (6, 1, 1.0),
            (1, 2, 1.0),
            (2, 3, 1.0),
            (3, 4, 1.0),
            (4, 1, -100.0),
        ]);
        let negative_cycle = ShortestPathGraph::new(&graph, 0.into(), Some(7));
    }
```
Add `println!("Loop found at step {} out of {}", steps - step, steps);` after `if pred == end_node {` to see when the cycle is found, in this case the output is `Loop found at step 5 out of 8`. The step at which the loop is found is always 5 even when the "tail" to zero becomes longer.